src/test/tests.rs

```diff
@@ -60,33 +59,44 @@ struct ServerOptions {
 /// * An `Arc`-and-`Mutex`-wrapped `MockCommandCreator` which the server will
 ///   use for all process creation.
 /// * The `JoinHandle` for the server thread.
-fn run_server_thread<T: Into<Option<ServerOptions>> + Send + 'static>(cache_dir: &Path, options: T) -> (u16, Sender<ServerMessage>, Arc<Mutex<MockCommandCreator>>, thread::JoinHandle<()>) {
+fn run_server_thread<T>(cache_dir: &Path, options: T)
+                        -> (u16, Sender<ServerMessage>, Arc<Mutex<MockCommandCreator>>, thread::JoinHandle<()>)
+    where T: Into<Option<ServerOptions>> + Send + 'static
+{
     let options = options.into();
+    let cache_dir = cache_dir.to_path_buf();
+
     // Create a server on a background thread, get some useful bits from it.
     let (tx, rx) = mpsc::channel();
-    let storage = Box::new(DiskCache::new(&cache_dir, options.as_ref().and_then(|o| o.cache_size.as_ref()).map(|s| *s).unwrap_or(usize::MAX)));
+    let (shutdown_tx, shutdown_rx) = oneshot::channel();
     let handle = thread::spawn(move || {
-        let (mut server, event_loop) = create_server::<Arc<Mutex<MockCommandCreator>>>(0, storage).unwrap();
-        assert!(server.port() > 0);
+        let srv = SccacheServer::new(0).unwrap();
+        let mut srv: SccacheServer<Arc<Mutex<MockCommandCreator>>> = srv;
+        assert!(srv.port() > 0);
+        let cache_size = options.as_ref()
+                                .and_then(|o| o.cache_size.as_ref())
+                                .map(|s| *s)
+                                .unwrap_or(usize::MAX);
         if let Some(options) = options {
             if let Some(timeout) = options.idle_timeout {
-                 server.set_idle_timeout(timeout);
+                 srv.set_idle_timeout(Duration::from_millis(timeout));
             }
         }
-        let port = server.port();
-        let sender = event_loop.channel();
-        let creator = server.command_creator();
-        tx.send((port, sender, creator)).unwrap();
-        run_server(server, event_loop).unwrap()
+        let storage = Arc::new(DiskCache::new(&cache_dir, cache_size, srv.pool()));
+        srv.set_storage(storage.clone());
+        let port = srv.port();
+        let creator = srv.command_creator().clone();
+        tx.send((port, creator)).unwrap();
+        srv.run(shutdown_rx).unwrap();
     });
-    let (port, sender, creator) = rx.recv().unwrap();
-    (port, sender, creator, handle)
+    let (port, creator) = rx.recv().unwrap();
+    (port, shutdown_tx, creator, handle)
 }
 
 #[test]
 fn test_server_shutdown() {
     let f = TestFixture::new();
-    let (port, _, _, child) = run_server_thread(&f.tempdir.path(), None);
```

Comment:
Oh this was actually super subtle IIRC in the sense that `let (a, _) = foo()` immediately destroys the second value. Destroying the second value here, `_sender`, would cause the event loop to exit immediately which wouldn't be desired.